File: app/src/main/java/com/facedetection/gaze/PerformanceMonitor.kt
package com.facedetection.gaze

import android.util.Log
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.math.roundToInt

/**
 * Performance monitor for tracking FPS, processing time, and other metrics.
 */
class PerformanceMonitor {
    
    companion object {
        private const val TAG = "PerformanceMonitor"
        private const val WINDOW_SIZE = 30 // Rolling window for averages
    }
    
    private val frameTimestamps = ConcurrentLinkedQueue<Long>()
    private val processingTimes = ConcurrentLinkedQueue<Long>()
    private var lastFrameTime = 0L
    private var frameCount = 0L
    private var totalProcessingTime = 0L
    
    data class PerformanceMetrics(
        val fps: Float,
        val avgProcessingTimeMs: Float,
        val maxProcessingTimeMs: Long,
        val droppedFrames: Int,
        val totalFrames: Long
    )
    
    /**
     * Record start of frame processing.
     */
    fun startFrame(): Long {
        return System.nanoTime()
    }
    
    /**
     * Record end of frame processing.
     */
    fun endFrame(startTime: Long) {
        val currentTime = System.nanoTime()
        val processingTime = (currentTime - startTime) / 1_000_000 // Convert to ms
        
        frameCount++
        totalProcessingTime += processingTime
        
        // Add to rolling window
        frameTimestamps.offer(currentTime)
        processingTimes.offer(processingTime)
        
        // Keep only recent data
        while (frameTimestamps.size > WINDOW_SIZE) {
            frameTimestamps.poll()
            processingTimes.poll()
        }
        
        lastFrameTime = currentTime
    }
    
    /**
     * Get current performance metrics.
     */
    fun getMetrics(): PerformanceMetrics {
        if (frameTimestamps.isEmpty()) {
            return PerformanceMetrics(0f, 0f, 0L, 0, frameCount)
        }
        
        // Calculate FPS
        val timestamps = frameTimestamps.toList()
        val fps = if (timestamps.size > 1) {
            val timeSpan = (timestamps.last() - timestamps.first()) / 1_000_000_000.0 // Convert to seconds
            if (timeSpan > 0) (timestamps.size / timeSpan).toFloat() else 0f
        } else {
            0f
        }
        
        // Calculate average processing time
        val times = processingTimes.toList()
        val avgProcessingTime = if (times.isNotEmpty()) {
            times.average().toFloat()
        } else {
            0f
        }
        
        // Find max processing time
        val maxProcessingTime = times.maxOrNull() ?: 0L
        
        // Estimate dropped frames (processing time > frame time)
        val targetFrameTime = 1000f / 30f // Assuming 30 FPS target
        val droppedFrames = times.count { it > targetFrameTime }
        
        return PerformanceMetrics(
            fps = fps,
            avgProcessingTimeMs = avgProcessingTime,
            maxProcessingTimeMs = maxProcessingTime,
            droppedFrames = droppedFrames,
            totalFrames = frameCount
        )
    }
    
    /**
     * Log current performance metrics.
     */
    fun logMetrics() {
        val metrics = getMetrics()
        Log.d(TAG, """
            Performance Metrics:
            FPS: ${metrics.fps.roundToInt()}
            Avg Processing Time: ${metrics.avgProcessingTimeMs.roundToInt()} ms
            Max Processing Time: ${metrics.maxProcessingTimeMs} ms
            Dropped Frames: ${metrics.droppedFrames}
            Total Frames: ${metrics.totalFrames}
        """.trimIndent())
    }
    
    /**
     * Reset all metrics.
     */
    fun reset() {
        frameTimestamps.clear()
        processingTimes.clear()
        lastFrameTime = 0L
        frameCount = 0L
        totalProcessingTime = 0L
    }
    
    /**
     * Check if performance is degraded.
     */
    fun isPerformanceDegraded(): Boolean {
        val metrics = getMetrics()
        return metrics.fps < 15f || metrics.avgProcessingTimeMs > 100f
    }
}

File: ios-sdk/Sources/GazePointSDK/PerformanceMonitor.swift
import Foundation

/// Performance monitor for tracking FPS, processing time, and other metrics
public class PerformanceMonitor {
    
    // MARK: - Types
    
    public struct PerformanceMetrics: Codable {
        public let fps: Float
        public let avgProcessingTimeMs: Float
        public let maxProcessingTimeMs: Float
        public let droppedFrames: Int
        public let totalFrames: Int64
        
        public init(fps: Float, avgProcessingTimeMs: Float, maxProcessingTimeMs: Float, droppedFrames: Int, totalFrames: Int64) {
            self.fps = fps
            self.avgProcessingTimeMs = avgProcessingTimeMs
            self.maxProcessingTimeMs = maxProcessingTimeMs
            self.droppedFrames = droppedFrames
            self.totalFrames = totalFrames
        }
    }
    
    // MARK: - Properties
    
    private let windowSize = 30 // Rolling window for averages
    private let targetFPS: Float = 30.0
    
    private var frameTimestamps: [TimeInterval] = []
    private var processingTimes: [TimeInterval] = []
    private var frameCount: Int64 = 0
    private var totalProcessingTime: TimeInterval = 0
    
    private let queue = DispatchQueue(label: "com.gazepoint.performancemonitor", attributes: .concurrent)
    
    // MARK: - Public Methods
    
    public init() {}
    
    /// Record start of frame processing
    public func startFrame() -> TimeInterval {
        return Date().timeIntervalSince1970
    }
    
    /// Record end of frame processing
    public func endFrame(startTime: TimeInterval) {
        let currentTime = Date().timeIntervalSince1970
        let processingTime = (currentTime - startTime) * 1000 // Convert to ms
        
        queue.async(flags: .barrier) { [weak self] in
            guard let self = self else { return }
            
            self.frameCount += 1
            self.totalProcessingTime += processingTime
            
            // Add to rolling window
            self.frameTimestamps.append(currentTime)
            self.processingTimes.append(processingTime)
            
            // Keep only recent data
            if self.frameTimestamps.count > self.windowSize {
                self.frameTimestamps.removeFirst()
                self.processingTimes.removeFirst()
            }
        }
    }
    
    /// Get current performance metrics
    public func getMetrics() -> PerformanceMetrics {
        return queue.sync {
            guard !frameTimestamps.isEmpty else {
                return PerformanceMetrics(
                    fps: 0,
                    avgProcessingTimeMs: 0,
                    maxProcessingTimeMs: 0,
                    droppedFrames: 0,
                    totalFrames: frameCount
                )
            }
            
            // Calculate FPS
            let fps: Float
            if frameTimestamps.count > 1 {
                let timeSpan = frameTimestamps.last! - frameTimestamps.first!
                fps = timeSpan > 0 ? Float(frameTimestamps.count) / Float(timeSpan) : 0
            } else {
                fps = 0
            }
            
            // Calculate average processing time
            let avgProcessingTime = processingTimes.isEmpty ? 0 : Float(processingTimes.reduce(0, +) / Double(processingTimes.count))
            
            // Find max processing time
            let maxProcessingTime = processingTimes.max() ?? 0
            
            // Estimate dropped frames
            let targetFrameTime = 1000.0 / Double(targetFPS)
            let droppedFrames = processingTimes.filter { $0 > targetFrameTime }.count
            
            return PerformanceMetrics(
                fps: fps,
                avgProcessingTimeMs: avgProcessingTime,
                maxProcessingTimeMs: Float(maxProcessingTime),
                droppedFrames: droppedFrames,
                totalFrames: frameCount
            )
        }
    }
    
    /// Log current performance metrics
    public func logMetrics() {
        let metrics = getMetrics()
        print("""
        Performance Metrics:
        FPS: \(Int(metrics.fps))
        Avg Processing Time: \(Int(metrics.avgProcessingTimeMs)) ms
        Max Processing Time: \(Int(metrics.maxProcessingTimeMs)) ms
        Dropped Frames: \(metrics.droppedFrames)
        Total Frames: \(metrics.totalFrames)
        """)
    }
    
    /// Reset all metrics
    public func reset() {
        queue.async(flags: .barrier) { [weak self] in
            self?.frameTimestamps.removeAll()
            self?.processingTimes.removeAll()
            self?.frameCount = 0
            self?.totalProcessingTime = 0
        }
    }
    
    /// Check if performance is degraded
    public func isPerformanceDegraded() -> Bool {
        let metrics = getMetrics()
        return metrics.fps < 15 || metrics.avgProcessingTimeMs > 100
    }
}
